### PR TITLE
Remove LINQ usages from product code

### PR DESCRIPTION
**Pull request by kasperk81**
cc @baywet

### CODE REVIEW

File: src/CborParseNode.cs [hunk 1]
```diff
@@ -238,9 +243,17 @@ internal CborParseNode(object? value)
...
-                return arrayValue.OfType<CborParseNode>().Select(x => x.GetObjectValue(factory));
+                List<T> result = new List<T>();
+                foreach(var item in arrayValue)
+                {
+                    if(item is CborParseNode node)
+                    {
+                        result.Add(node.GetObjectValue(factory));
+                    }
+                }
+                return result;
```
Comment (baywet) [inline src/CborParseNode.cs]:
```suggestion
                foreach(var item in arrayValue)
                {
                    if(item is CborParseNode node)
                    {
                        yield return node.GetObjectValue(factory);
                    }
                }
```

File: src/CborParseNode.cs [hunk 2]
```diff
@@ -321,12 +334,30 @@ private static T GetItemValue<T>(Type genericType, CborParseNode item)
...
             var genericType = typeof(T);
+            List<T> result = new List<T>();
+
             if(value is object?[] arrayValue1)
-                return arrayValue1.OfType<CborParseNode>().Select(x => GetItemValue<T>(genericType, x));
+            {
+                foreach(var item in arrayValue1)
+                {
+                    if(item is CborParseNode node)
+                    {
```
Comment (baywet) [inline src/CborParseNode.cs]: switch to yield returns as well

File: src/CborParseNode.cs [hunk 3]
```diff
@@ -403,32 +428,58 @@ public IEnumerable<T> GetCollectionOfPrimitiveValues<T>()
...
+            }
+            return newArray;
+        }
+
         /// <summary>
         /// Get the child node of the specified identifier
         /// </summary>
         /// <param name="identifier">The identifier of the child node</param>
         /// <returns>An instance of <see cref="IParseNode"/></returns>
-        public IParseNode? GetChildNode(string identifier)
```
Comment (andrueastman) [inline src/CborParseNode.cs:420]: Looks like `GetChildNode` got dropped which is a contract defined by the parent interace.

File: src/CborParseNode.cs [hunk 4]
```diff
@@ -322,11 +333,22 @@ public IEnumerable<T> GetCollectionOfPrimitiveValues<T>()
...
+            IEnumerable<T> GetValuesFromArray<T>(Type genericType, object?[] array)
+            {
+                foreach (var item in array)
+                {
+                    if (item is CborParseNode node)
+                    {
+                        yield return GetItemValue<T>(genericType, node);
+                    }
+                }
+            }
```
Comment (andrueastman) [inline src/CborParseNode.cs]: `T` matches outer function `T`
```suggestion
            static IEnumerable<TItem> GetValuesFromArray<TItem>(Type genericType, object?[] array)
            {
                foreach (var item in array)
                {
                    if (item is CborParseNode node)
                    {
                        yield return GetItemValue<TItem>(genericType, node);
                    }
                }
            }
```

File: src/CborParseNode.cs (hunk 3 above)
Comment (kasperk81) [reply src/CborParseNode.cs:420]: copy & paste mistake. fixed

File: src/CborSerializationWriter.cs [hunk 5]
```diff
@@ -250,9 +259,18 @@ public void WriteNullValue(string? key)
...
+                //empty array is meaningful
                 if(!string.IsNullOrEmpty(key)) writer.WriteTextString(key!);
-                writer.WriteStartArray(values.Count());
+
+                // Count elements without using LINQ
+                int count = 0;
+                foreach(var _ in values)
+                {
+                    count++;
+                }
```
Comment (kasperk81) [inline src/CborSerializationWriter.cs]:
```suggestion
                foreach(var _ in values) count++;
```

File: src/CborSerializationWriter.cs [hunk 6]
```diff
@@ -267,14 +282,24 @@ public void WriteCollectionOfPrimitiveValues<T>(string? key, IEnumerable<T>? val
...
```
Comment (kasperk81) [inline src/CborSerializationWriter.cs]:
```suggestion
                foreach(var _ in values) count++;
```

File: src/CborSerializationWriter.cs [hunk 7]
```diff
@@ -283,23 +305,33 @@ public void WriteCollectionOfPrimitiveValues<T>(string? key, IEnumerable<T>? val
...
```
Comment (kasperk81) [inline src/CborSerializationWriter.cs]:
```suggestion
                foreach(var _ in values) count++;
```

**Comment by andrueastman** [comment]
Any chance you can doulbe check the failing test at https://github.com/microsoft/kiota-serialization-cbor-dotnet/actions/runs/9251381507/job/25448819544?pr=76

**Review by andrueastman — APPROVED** [review]

**Pull request merged**